### PR TITLE
Cache docker image information

### LINKED ISSUE #3152: Cache docker layer info for percentage calculation
Here's a simple idea that I think might significantly improve the accuracy of our percentage calculations. This would replace #3105 .

The idea is to cache the information about layers and their sizes. This way we'll be inaccurate on the first pull of an image, but all subsequent pulls should be much more accurate. We compute and cache `progressDetailsInit`, which is an object that we will use to initialize `progressDetails` (that is, it has all the totals, but all the current values set to zero).

* [ ] Replace https://github.com/PrairieLearn/PrairieLearn/blob/03f3c1c6882ad27601d2ecd13534f645c2115368/workspace_host/interface.js#L886 with an attempt to get the cached information:
```
let progressDetails = await cache.getAsync(`workspaceProgressInit:${workspace_image}`) || {};
let progressDetailsInit = {};
```

* [ ] Add the update of the initial value at https://github.com/PrairieLearn/PrairieLearn/blob/03f3c1c6882ad27601d2ecd13534f645c2115368/workspace_host/interface.js#L903
```
progressDetailsInit[key] = {...output.progressDetail, current: 0};
```

* [ ] Once we've got all the info we should set the cached value at https://github.com/PrairieLearn/PrairieLearn/blob/03f3c1c6882ad27601d2ecd13534f645c2115368/workspace_host/interface.js#L894
```
cache.set(`workspaceProgressInit:${workspace_image}`, progressDetailsInit);
```

### PR DESCRIPTION
**Pull request by mylesw27**
This PR intends to provide more accurate percentage calculation when loading workspaces. This is accomplished by caching information about docker layers and their sizes after the first pull of a workspace image. This will allow us to compute more accurate percentages based on historical data. Closes #3152 .

### CODE REVIEW

File: apps/workspace-host/src/lib/config.ts [hunk 1]
```diff
@@ -13,6 +13,8 @@ const ConfigSchema = z.object({
   postgresqlHost: z.string().default('localhost'),
   postgresqlPoolSize: z.number().default(100),
   postgresqlIdleTimeoutMillis: z.number().default(30_000),
+  cacheType: z.enum(['none', 'redis', 'memory']).default('none'),
+  cacheKeyPrefix: z.string().default('prairielearn-cache:'),
```
Comment (nwalters512) [inline apps/workspace-host/src/lib/config.ts:17]: Let's use a different default namespace here.

File: apps/workspace-host/src/interface.js [hunk 2]
```diff
@@ -26,6 +26,7 @@ import { logger } from '@prairielearn/logger';
 import * as sqldb from '@prairielearn/postgres';
 import * as Sentry from '@prairielearn/sentry';
 import * as workspaceUtils from '@prairielearn/workspace-utils';
+import { cache } from '@prairielearn/cache';
```
Comment (nwalters512) [inline apps/workspace-host/src/interface.js:29]: You should add this package to `apps/workspace-host/package.json`.

File: apps/workspace-host/src/interface.js [hunk 3]
```diff
@@ -251,6 +252,13 @@ async
         await dockerAttemptKillAndRemove(docker.getContainer(container.Id));
       }
     },
+    async () => {
+      cache.init({
```
Comment (nwalters512) [inline apps/workspace-host/src/interface.js:256]: This function is async; you need to await the returned promise.

File: apps/workspace-host/src/interface.js [hunk 4]
```diff
@@ -628,6 +637,7 @@ async function _pullImage(workspace) {
         }
 
         const toDatabase = false;
+        cache.set(`workspaceProgressInit:${workspace_image}`, progressDetailsInit, 1000 * 60 * 60);
```
Comment (nwalters512) [inline apps/workspace-host/src/interface.js:640]: @mwest1066 you wanted to make the expiration here really long, right?

File: apps/workspace-host/src/interface.js [hunk 5]
```diff
@@ -603,7 +611,8 @@ async function _pullImage(workspace) {
   // will tend to go faster at the start (when we only know
   // about a few layers) and slow down at the end (when we
   // know about all layers).
-  let progressDetails = {};
+  let progressDetails = (await cache.get(`workspaceProgressInit:${workspace_image}`)) || {};
```
Comment (nwalters512) [inline apps/workspace-host/src/interface.js:614]: Can you add a comment explaining why we're caching this?

File: apps/workspace-host/src/interface.js (hunk 4 above)
Comment (nwalters512) [inline apps/workspace-host/src/interface.js:640]: `cache.set` is an async function; you should ensure errors are properly handled. See `workspaceUtils.updateWorkspaceMessage(...)` on the next line for how to do that.

Comment (nwalters512) [reply apps/workspace-host/src/interface.js:640]: As Myles pointed out, `cache.set` is _not_ in fact an async function, it doesn't return anything at all!

File: apps/workspace-host/src/interface.js (hunk 2 above)
Comment (mylesw27) [reply apps/workspace-host/src/interface.js:29]: Added to this package to `apps/workspace-host/package.json` in [eae6537](https://github.com/PrairieLearn/PrairieLearn/pull/9408/commits/eae65374a2f77e98f4937bcd4537406b04524243).

File: apps/workspace-host/src/interface.js (hunk 3 above)
Comment (mylesw27) [reply apps/workspace-host/src/interface.js]: Updated to await this promis in [eae6537](https://github.com/PrairieLearn/PrairieLearn/pull/9408/commits/eae65374a2f77e98f4937bcd4537406b04524243).

File: apps/workspace-host/src/interface.js (hunk 4 above)
Comment (mylesw27) [reply apps/workspace-host/src/interface.js]: Updated this to expire in 24 hours. Let me know if you want me to adjust here!

File: apps/workspace-host/src/interface.js [hunk 6]
```diff
@@ -628,6 +641,11 @@ async function _pullImage(workspace) {
         }
 
         const toDatabase = false;
+        cache.set(
+          `workspaceProgressInit:${workspace_image}`,
+          progressDetailsInit,
+          1000 * 60 * 60 * 24, // 24 hours
```
Comment (mwest1066) [inline apps/workspace-host/src/interface.js:647]: I think we should cache for longer than 24 hours -- maybe a month?

[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/workspace-host/src/interface.js]: Per our discussion in the dev meeting, I have updated this to a month. There is not much data being cached here and there is not much risk involved in keeping this around longer.

Comment (mylesw27) [reply apps/workspace-host/src/interface.js]: I have updated the cache expiration to 30 days in 2eefe46.

**Review by nwalters512 — APPROVED** [review]
A note for posterity: the caching strategy used here will result in slightly strange behavior when the underlying image changes.

Consider an image with the following layers:

- A (100 bytes)
- B (200 bytes)
- C (300 bytes)

Now, we push a new version of the image with the following layers:

- A (100 bytes)
- D (600 bytes)
- E (1000 bytes)

We'll still use the cached layer sizes from the first version as a starting point, but we'll add layers D and E as we go. This will result in us finishing after loading 1700 bytes, but the total number of bytes for the percentage calculation will be 2200 bytes, meaning the container will start when the progress bar displays 77% or so.

In practice, this is entirely OK, no one will complain about something finishing early. My concern had been that we'd have the opposite problem, where a new image would somehow result in the progress bar staying at 99% for a long time. I think we're unlikely to run into that scenario unless an image has lots of layers and one of the final layers grows significantly larger. Regardless, this change is an obvious improvement over the status quo, and images shouldn't change too often, so this concern won't apply most of the time anyways!

**Pull request merged**